data/abi/abi.go

```diff
@@ -0,0 +1,1164 @@
+		return 0, fmt.Errorf("value type unmatch or size too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint32(bigIntForm.Uint64()), nil
+}
+
+// GetUint64 tries to retrieve an uint64 from an ABI Value.
+func (v Value) GetUint64() (uint64, error) {
+	if v.ABIType.enumIndex != Uint || v.ABIType.size > 64 {
+		return 0, fmt.Errorf("value type unmatch or size too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return bigIntForm.Uint64(), nil
+}
+
+// GetUint tries to retrieve an big uint from an ABI Value.
+func (v Value) GetUint() (*big.Int, error) {
+	if v.ABIType.enumIndex != Uint {
+		return nil, fmt.Errorf("value type unmatch")
+	}
+	bigIntForm := v.value.(*big.Int)
+	sizeThreshold := big.NewInt(0).Lsh(big.NewInt(1), uint(v.ABIType.size))
+	if sizeThreshold.Cmp(bigIntForm) <= 0 {
+		return nil, fmt.Errorf("value is larger than uint size")
+	}
+	return bigIntForm, nil
+}
+
+// GetUfixed tries to retrieve an big rational number from an ABI Value.
+func (v Value) GetUfixed() (*big.Int, error) {
+	if v.ABIType.enumIndex != Ufixed {
+		return nil, fmt.Errorf("value type unmatch, should be ufixed")
+	}
+	bigIntForm := v.value.(*big.Int)
+	sizeThreshold := big.NewInt(0).Lsh(big.NewInt(1), uint(v.ABIType.size))
+	if sizeThreshold.Cmp(bigIntForm) <= 0 {
+		return nil, fmt.Errorf("value is larger than ufixed size")
+	}
+	return bigIntForm, nil
+}
+
+// GetString tries to retrieve a string from ABI Value.
+func (v Value) GetString() (string, error) {
+	if v.ABIType.enumIndex != String {
+		return "", fmt.Errorf("value type unmatch, should be ufixed")
+	}
+	stringForm := v.value.(string)
+	return stringForm, nil
+}
+
+// GetByte tries to retrieve a byte from ABI Value.
+func (v Value) GetByte() (byte, error) {
+	if v.ABIType.enumIndex != Byte {
+		return byte(0), fmt.Errorf("value type unmatch, should be bytes")
+	}
+	bytesForm := v.value.(byte)
+	return bytesForm, nil
+}
+
+// GetAddress tries to retrieve a [32]byte array from ABI Value.
+func (v Value) GetAddress() ([32]byte, error) {
+	if v.ABIType.enumIndex != Address {
+		return [32]byte{}, fmt.Errorf("value type unmatch, should be address")
+	}
+	addressForm := v.value.([32]byte)
+	return addressForm, nil
+}
+
+// GetValueByIndex retrieve value element by the index passed in
+func (v Value) GetValueByIndex(index uint16) (Value, error) {
+	switch v.ABIType.enumIndex {
+	case ArrayDynamic:
+		elements := v.value.([]Value)
+		if len(elements) <= int(index) {
+			return Value{}, fmt.Errorf("cannot get element: index out of scope")
+		}
+		return elements[index], nil
+	case ArrayStatic, Tuple:
+		elements := v.value.([]Value)
+		if v.ABIType.staticLength <= index {
+			return Value{}, fmt.Errorf("cannot get element: index out of scope")
+		}
+		return elements[index], nil
+	default:
+		return Value{}, fmt.Errorf("cannot get value by index for non array-like type")
+	}
+}
```

Comment:
I merged `Get` method for tuple and arrays into `GetValueByIndex`. See if it needs to be simplified.